Add simple feature flag system

oneiros opened the pull request:
In #34031 I introduced a very basic, singular feature flag. In the future we would like to have more and be able to introduce and use them quickly.

That is why @renchap suggested a slightly more dynamic version that can be easily re-used.

It allows specifying enabled features as a comma separated list in the `EXPERIMENTAL_FEATURES` environment variable:

```sh
EXPERIMENTAL_FEATURES=fasp,other
```

This can then be queried like this:

```ruby
Mastodon::Feature.fasp_enabled? # => true
Mastodon::Feature.other_enabled? # => true

Mastodon::Feature.unknown_enabled? # => false
```

oneiros committed:
Add simple feature flag system

Allows specifying enabled features as a comma separated list
in `ENV['EXPERIMENTAL_FEATURES']`.

Can be queries as `Mastodon::Feature.<name>_enabled?`.

mjankowski commented on lib/mastodon/feature.rb:
```diff
@@ -0,0 +1,26 @@
+  class << self
+    def enabled_features
+      @enabled_features ||=
+        (ENV['EXPERIMENTAL_FEATURES'] || '').split(',').map(&:strip)
```
Is the idea that this is loaded once at launch time and would require a restart to reload the features? (ie, no dynamic changing w/out restart) - if so, could shove this into one of the `config_for` namespaces?

Separately, maybe `ENV.fetch...`?

oneiros replied:
> Separately, maybe ENV.fetch...?

Definitely! Thanks a lot.

> Is the idea that this is loaded once at launch time and would require a restart to reload the features? (ie, no dynamic changing w/out restart) 

Yes, that is all we need for now.

> if so, could shove this into one of the config_for namespaces?

I am not sure. I should maybe add that one of the requirements was that toggling of multiple features should be possible via a single environment variable. I believe it would be awkward to implement this as ERB within a yaml file.

Or what is it you are proposing here exactly?

mjankowski replied:
> I am not sure. I should maybe add that one of the requirements was that toggling of multiple features should be possible via a single environment variable. I believe it would be awkward to implement this as ERB within a yaml file.

I agree that once you start to cross the line from "just loading some env vars" into "processing strings and running ruby code" in a yaml file it starts to feel awkward. I'm not sure where that line is exactly.

So far we mostly just have the env var loading in the yml - https://github.com/mastodon/mastodon/blob/main/config/mastodon.yml - and I think most scenarios that need additional parsing/processing do it in private methods where they are used?

One side goal of the `config_for` migration was to eventually enable https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsenvironmentvariableaccess and enforce the "configure at load time" practice.

mjankowski replied:
Yup, exactly that - just the bare minimum of nudging the env load into `config_for`, and leaving any parsing/formatting/etc in the classes.

oneiros replied:
I gave it a try and it seems to work.

ClearlyClaire commented on lib/mastodon/feature.rb:
```diff
@@ -0,0 +1,26 @@
+
+    def method_missing(name)
+      if respond_to_missing?(name)
+        feature = name.to_s.sub(/_enabled\?$/, '')
```
More readable I think and a tiny bit faster:
```suggestion
        feature = name.to_s.delete_suffix('_enabled?')
```

oneiros replied:
Thanks, that is brilliant. I keep forgetting these methods exist. I mean they have only been introduced ... *checks notes* ... 7 years ago :man_facepalming: :older_man:

ClearlyClaire reviewed (dismissed):
Looks good to me with a little nitpick. Using `config_for` as @mjankowski suggested might also make sense.

oneiros committed:
Use much better method

Co-authored-by: Claire <claire.github-309c@sitedethib.com>

oneiros committed:
Use `ENV.fetch` as suggested in review

oneiros committed:
Move access to ENV into config file

The end goal being to use this cop here in the future:
https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsenvironmentvariableaccess
